pkg/did/client.go

```diff
@@ -145,8 +210,162 @@ func unwrapPubKeyJWK(key PublicKey) (*PublicKey, error) { // nolint: gocritic
+	nextUpdateKey, err := pubkey.GetPublicKeyJWK(updateDIDOpts.nextUpdatePublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get next update key : %s", err)
+	}
+
+	nextUpdateCommitment, err := commitment.Calculate(nextUpdateKey, sha2_256, sha256)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, updateKey, err := getSigner(updateDIDOpts.signingkey)
+	if err != nil {
+		return nil, err
+	}
+
+	patches, err := createUpdatePatches(updateDIDOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	didSuffix, err := getUniqueSuffix(did)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.NewUpdateRequest(&client.UpdateRequestInfo{
+		DidSuffix:        didSuffix,
+		UpdateCommitment: nextUpdateCommitment,
+		UpdateKey:        updateKey,
+		Patches:          patches,
+		MultihashCode:    sha2_256,
+		Signer:           signer,
+	})
+}
+
+func getSigner(signingkey crypto.PrivateKey) (client.Signer, *jws.JWK, error) {
+	switch key := signingkey.(type) {
+	case *ecdsa.PrivateKey:
+		updateKey, err := pubkey.GetPublicKeyJWK(key.Public())
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if key.Curve.Params().Name != elliptic.P256().Params().Name {
+			return nil, nil, fmt.Errorf("ec cruve %s key not supported", elliptic.P256().Params().Name)
+		}
+
+		return ecsigner.New(key, "ES256", "update-kid"), updateKey, nil
+	case ed25519.PrivateKey:
+		updateKey, err := pubkey.GetPublicKeyJWK(key.Public())
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return edsigner.New(key, "EdDSA", "key-1"), updateKey, nil
+	default:
+		return nil, nil, fmt.Errorf("key not supported")
+	}
+}
+
+func getUniqueSuffix(id string) (string, error) {
+	p := strings.LastIndex(id, ":")
+	if p == -1 {
+		return "", fmt.Errorf("unique suffix not provided in URL [%s]", id)
+	}
+
+	return id[p+1:], nil
+}
+
+func createUpdatePatches(updateDIDOpts *UpdateDIDOpts) ([]patch.Patch, error) {
+	var patches []patch.Patch
+
+	if len(updateDIDOpts.removePublicKeys) != 0 {
+		p, err := createRemovePublicKeysPatch(updateDIDOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		patches = append(patches, p)
+	}
+
+	if len(updateDIDOpts.removeServices) != 0 {
+		p, err := createRemoveServicesPatch(updateDIDOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		patches = append(patches, p)
+	}
+
+	if len(updateDIDOpts.addServices) != 0 {
+		p, err := createAddServicesPatch(updateDIDOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		patches = append(patches, p)
+	}
+
+	if len(updateDIDOpts.addPublicKeys) != 0 {
+		p, err := createAddPublicKeysPatch(updateDIDOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		patches = append(patches, p)
+	}
+
+	return patches, nil
+}
+
+func createRemovePublicKeysPatch(updateDIDOpts *UpdateDIDOpts) (patch.Patch, error) {
+	removePubKeys, err := json.Marshal(updateDIDOpts.removePublicKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	return patch.NewRemovePublicKeysPatch(string(removePubKeys))
```

Comment:
I am wondering if I should have accepted bytes instead of string in New() - nothing to do with your code, just thinking ...